





# Data download

For each species identified in the bladder by Thomas-White, the 16S rRNA
gene sequence of the corresponding type strain was downloaded from the
Silva v132 release on 4/27/2019. A type strain is the sequence of the
cultured isolate that was subject to the metabolic, genotypic and
phenotypic evaluations taken to define the bacterial species\[1\], and
is the agreed bacterial organism to which the taxonomic name is
referring.

## Synonyms of species

Species names have changed in response to advances in bacterial
systematics. All currently known species synonyms were downloaded from
the Prokaryotic Nomenclature Up-to-Date\[2\] (PNU) website on 1/5/2020.
PNU includes information down to the strain level, but these entries
were consolidated to the species level. For example, entries like
*Enterobacter cloacae* and *Enterobacter cloacae dissolvens* are treated
as synonyms of *Enterobacter cloacae*.

## Silva search parameters

Sequences were searched using the “\[T\]” filter setting, and sequences
longer than 1450nt with alignment and pintail quality scores greater
than 95% were selected. For the species that had no hits, the taxonomic
synonym was used as the search query (if the known synonym was
available). The species identified as *Corynebacterium sp* had no type
strain available, and was excluded from the query set. Species and
accession numbers are listed
below.

| species                                       | type strain accession | notes                                      |
| --------------------------------------------- | --------------------- | ------------------------------------------ |
| Actinobaculum schaalii                        | FJ960443              | as Actinotignum schaali                    |
| Actinomyces naeslundii                        | AB618790              |                                            |
| Actinomyces neuii                             | AM084229              | Actinomyces neuii subsp. anitratus         |
| Actinomyces odontolyticus                     | AB818950              |                                            |
| Actinomyces turicensis                        | X78720                |                                            |
| Actinomyces urogenitalis                      | ACFH01000038          | Actinomyces urogenitalis DSM 15434         |
| Aerococcus christensenii                      | Y17005                |                                            |
| Aerococcus sanguinicola                       | AJ276512              |                                            |
| Aerococcus urinae                             | M77819                |                                            |
| Alloscardovia omnicolens                      | AM419460              |                                            |
| Anaerococcus octavius                         | Y07841                |                                            |
| Bacillus idriensis                            | AY904033              |                                            |
| Bacillus infantis                             | AY904032              |                                            |
| Bifidobacterium bifidum                       | S83624                |                                            |
| Bifidobacterium breve                         | AB006658              |                                            |
| Bifidobacterium longum                        | AP010889              | Bifidobacterium longum subsp. infa…        |
| Brevibacterium massiliense                    | EU086793              | Brevibacterium ravenspurgense              |
| Campylobacter ureolyticus                     | ARGD01000016          | Campylobacter ureolyticus DSM 20703        |
| Corynebacterium amycolatum                    | X84244                |                                            |
| Corynebacterium aurimucosum                   | AJ309207              |                                            |
| Corynebacterium coyleae                       | X96497                |                                            |
| Corynebacterium matruchotii                   | ACSH02000003          | Corynebacterium matruchotii ATCC 1…        |
| Corynebacterium pyruviciproducens             | ATBY01000001          |                                            |
| Corynebacterium riegelii                      | Y14651                |                                            |
| Corynebacterium sp.                           | .                     | no type spec                               |
| Corynebacterium tuscaniense                   | AY677186              |                                            |
| Dermabacter hominis                           | X91034                |                                            |
| Enterobacter asburiae                         | AB004744              |                                            |
| Enterobacter cloacae                          | CP001918              | Enterobacter cloacae subsp. cloaca…        |
| Enterococcus faecalis                         | AB012212              |                                            |
| Escherichia coli                              | EU014689              |                                            |
| Facklamia hominis                             | AGZD01000007          | Facklamia hominis CCUG 36813               |
| Facklamia ignava                              | AGZE01000005          | Facklamia ignava CCUG 37419                |
| Gardnerella vaginalis                         | M58744                | Gardnerella vaginalis ATCC 14018           |
| Globicatella sanguinis                        | AB680901              | Globicatella sanguinis                     |
| Gordonia terrae                               | AB920570              |                                            |
| Klebsiella pneumoniae                         | X87276                |                                            |
| Kocuria rhizophila                            | Y16264                |                                            |
| Kytococcus schroeteri                         | AJ297722              |                                            |
| Lactobacillus crispatus                       | AB008206              |                                            |
| Lactobacillus delbrueckii                     | AB007908              |                                            |
| Lactobacillus fermentum                       | AJ575812              |                                            |
| Lactobacillus gasseri                         | AB008209              |                                            |
| Lactobacillus iners                           | HE573916              |                                            |
| Lactobacillus jensenii                        | AF243176              |                                            |
| Lactobacillus johnsonii                       | HE573915              |                                            |
| Lactobacillus pontis                          | X76329                |                                            |
| Lactobacillus rhamnosus                       | AB008211              |                                            |
| Micrococcus luteus                            | CP001628              | Micrococcus luteus NCTC 2665               |
| Moraxella osloensis                           | FR726160              |                                            |
| Morganella morganii                           | AJ301681              | Morganella morganii subsp. morganii        |
| Neisseria macacae                             | HF558383              |                                            |
| Neisseria perflava                            | HF558366              |                                            |
| Neisseria subflava                            | AJ239291              |                                            |
| Oligella urethralis                           | AF133538              |                                            |
| Propionibacterium acnes                       | CP006032              | Propionibacterium acnes hdn-1 no type spec |
| Propionibacterium avidum                      | AGBA01000019          | as Cutibacterium avidum ATCC 25577         |
| Proteus mirabilis                             | AJ301682              |                                            |
| Pseudomonas aeruginosa                        | HE978271              |                                            |
| Rothia dentocariosa                           | CP002280              | Rothia dentocariosa ATCC 17931             |
| Rothia mucilaginosa                           | X87758                |                                            |
| Staphylococcus epidermidis                    | AB681292              |                                            |
| Staphylococcus hominis                        | X66101                | Staphylococcus hominis subsp. homi… 1      |
| Staphylococcus pettenkoferi                   | AF322002              |                                            |
| Staphylococcus saprophyticus                  | AP008934              | Staphylococcus saprophyticus subsp…        |
| Staphylococcus simulans                       | D83373                |                                            |
| Staphylococcus warneri                        | L37603                |                                            |
| Streptococcus agalactiae                      | AB023574              |                                            |
| Streptococcus anginosus                       | AFIM01000033          |                                            |
| Streptococcus equinus                         | AB680295              |                                            |
| Streptococcus gordonii                        | AF003931              |                                            |
| Streptococcus mitis                           | AF003929              |                                            |
| Streptococcus oralis                          | ADMV01000001          |                                            |
| Streptococcus parasanguinis                   | AF003933              |                                            |
| Streptococcus salivarius                      | AY188352              | Streptococcus salivarius subsp. sa…        |
| Streptococcus sanguinis                       | AF003928              |                                            |
| Trueperella bernardiae only one seq available | X79224                |                                            |
| Varibaculum cambriense                        | AJ428402              |                                            |
| Veillonella parvula                           | AB538437              |                                            |

## pare down results

All these sequences were downloaded and saved in one multi-record FASTA
file named arb-silva.de\_2019-11-10\_id739827.fasta. However, this file
also includes several other entries of each species. For example, in
addition to the desired type strain *Streptococcus equinus* AB680295,
Silva included an additional 6 entries in the download (AB002482.1.1457,
AF104116.1.356, AEVB01000043.3380.4915, AF429764.1.1430, AF429765.1.1463
and AJ301607.1.1433). We used the script `cull_downloads.py` to comb
through the Silva download and write the desired type strains to a new
file
    `ktw_16s_type_2019-11-25_34.fna`"\`.

    ~>python cull_downloads.py -i arb-silva.de_2019-11-10_id739827_tax_silva.fasta
    writing to ktw_16s_type_2019-11-25_34.fna

# Alignment

The sequences in the `ktw_16s_type_2019-11-25_34.fna` file is then used
to generate a multi-sequence alignement (MSA) of the 16S rRNA gene. We
used the program T-Coffee\[3\], an alignment program that generates
accurate multisequence alignments even when the sequences are from
unrelated species\[4\]. The program was run with default settings.

    ~$ t_coffee ktw_16s_type_2019-11-25_34.fna -outorder=aligned

The output of the program is a guide tree in Newick format (.dnd), an
html file of the MSA in a colorized version, and the actual alignment in
Clustal format (.aln). The Clustal file will be used to extract the
predicted amplicons.

# Extract amplicons

The MSA contains the 16S rRNA sequence of the *E. coli* type strain
EU014689, and this will be used as a reference to map the 9 known
variable regions onto the remaining 77 species from the Thomas-White
dataset. The positions of the variable regions were confirmed through
Sliding Window Analysis (see the sliding\_window folder).

| variable region | *E. coli* start | *E. coli* stop |  |  |
| --------------- | --------------- | -------------- |  |  |
| V1              | 69              | 99             |  |  |
| V2              | 137             | 242            |  |  |
| V3              | 433             | 497            |  |  |
| V4              | 576             | 682            |  |  |
| V5              | 822             | 879            |  |  |
| V6              | 986             | 1043           |  |  |
| V7              | 1117            | 1173           |  |  |
| V8              | 1243            | 1294           |  |  |
| V9              | 1435            | 1465           |  |  |

The python script `extract_16s_vr.py` is used to extract the predicted
amplicons from the MSA based on the 5’ and 3’ annealing sites of the
primer set that would be used in PCR. In the example below, the script
takes three arguments: the path to the MSA
(ktw\_type\_16s\_2019-11-25/aln/ktw\_16s\_type\_2019-11-25\_34.aln), the
name of the file that will be written (cap\_tpstr), and the start and
stop coordinates (516,808). The output is a multi-record FASTA file of
the predicted amplicon sequence, for each species in the
    MSA.

    ~>python extract_16s_vr.py -i ktw_type_16s_2019-11-25/aln/ktw_16s_type_2019-11-25_34.aln -o cap_tpstr -e 516,808
    writing to ../processed_data/extracted_16s_regions_2019-11-26/extracted_cap_tpstr_2019-11-26_1550.fna

There are two important notes. The first is that you’ll need to
determine where a primer set’s annealing sites correspond to the
sequence of the *E. coli* type strain EU014689. We used the program
UGENE\[5\] to search the MSA for the oligo sequence of the primer and
transposed to *E. coli*, but there are other methods that will work. The
second is that this script producs *predicted* amplicons. It doesn’t
check if the primers will actually anneal to a specific species’ 16S
rRNA gene sequence. That’s a feature that will probably be added later.

1.  Rainey FA. How to Describe New Species of Prokaryotes. In: Methods
    in Microbiology \[Internet\]. Elsevier; 2011 \[cited 2019 Nov 4\].
    p. 7–14. Available from:
    <https://linkinghub.elsevier.com/retrieve/pii/B9780123877307000024>

2.  Parte AC. LPSN – List of Prokaryotic names with Standing in
    Nomenclature (bacterio.net), 20 years on. Int J Syst Evol Microbiol.
    2018 Jun 1;68(6):1825–9.

3.  Notredame C, Higgins DG, Heringa J. T-coffee: a novel method for
    fast and accurate multiple sequence alignment 1 1Edited by J.
    Thornton. Journal of Molecular Biology. 2000 Sep;302(1):205–17.

4.  Edgar RC, Batzoglou S. Multiple sequence alignment. Current Opinion
    in Structural Biology. 2006 Jun;16(3):368–73.

5.  Okonechnikov K, Golosova O, Fursov M. Unipro UGENE: a unified
    bioinformatics toolkit. Bioinformatics. 2012 Apr;28(8):1166–7.
