---
title: "Extracting predicted amplicons"
author: "Carter Hoffman"
date: " `r Sys.Date()` "

header-includes:
   - \usepackage{amsmath,amssymb}
   - \usepackage[bitstream-charter]{mathdesign}
   - \usepackage[T1]{fontenc}
   
output: 
  md_document:
    variant: gfm

---
```{r, eval=FALSE, echo=FALSE}
output: 
  pdf_document:
      toc: true
      toc_depth: 2
      df_print: kable
    
output: 
  html_document:
    toc: true
    toc_depth: 4
    toc_float: true
    theme: readable
    highlight: tango

# knit directly to md for github
output:
  md_document:
    variant: gfm
```


```{r echo=FALSE}
knitr::opts_chunk$set(tidy = FALSE, warning = FALSE, message = FALSE, max.print = 5)
```

```{r echo=FALSE}
# load libraries
library(tidyverse)
library(patchwork)
library(wesanderson)
```

# Extracting predicted amplicons

## Data download

For each species identified in the bladder by Thomas-White, the 16S rRNA gene sequence of the corresponding type strain was downloaded from the Silva v132 release on 4/27/2019. A type strain is the sequence of the cultured isolate that was subject to the metabolic, genotypic and phenotypic evaluations taken to define the bacterial species[^3], and is the agreed bacterial organism to which the taxonomic name is referring. 

[^3]: Rainey FA. How to Describe New Species of Prokaryotes. In: Methods in Microbiology [Internet]. Elsevier; 2011 [cited 2019 Nov 4]. p. 7–14. Available from: https://linkinghub.elsevier.com/retrieve/pii/B9780123877307000024

### Synonyms of species

Species names have changed in response to advances in bacterial systematics. All currently known species synonyms were downloaded from the Prokaryotic Nomenclature Up-to-Date[^2] (PNU) website on 1/5/2020. PNU includes information down to the strain level, but these entries were consolidated to the species level. For example, entries like _Enterobacter cloacae_ and _Enterobacter cloacae dissolvens_ are treated as synonyms of _Enterobacter cloacae_. 

[^2]: Parte AC. LPSN – List of Prokaryotic names with Standing in Nomenclature (bacterio.net), 20 years on. Int J Syst Evol Microbiol. 2018 Jun 1;68(6):1825–9. 

### Silva search parameters

Sequences were searched using the “[T]” filter setting, and sequences longer than 1450nt with alignment and pintail quality scores greater than 95% were selected. For the species that had no hits, the taxonomic synonym was used as the search query (if the known synonym was available). The species identified as _Corynebacterium sp_ had no type strain available, and was excluded from the query set. Species and accession numbers are listed below.


| species | type strain accession | notes|
|-----|-----|-----|
| Actinobaculum schaalii                |FJ960443 |as Actinotignum schaali|
| Actinomyces naeslundii                |AB618790||
| Actinomyces neuii                     |AM084229 |	Actinomyces neuii subsp. anitratus|
| Actinomyces odontolyticus             |AB818950||
| Actinomyces turicensis                |X78720||
| Actinomyces urogenitalis              |ACFH01000038 |	Actinomyces urogenitalis DSM 15434|
| Aerococcus christensenii              |Y17005||
| Aerococcus sanguinicola               |AJ276512||
| Aerococcus urinae                     |M77819||
| Alloscardovia omnicolens              |AM419460||
| Anaerococcus octavius                |Y07841||
| Bacillus idriensis                    |AY904033||
| Bacillus infantis                     |AY904032||
| Bifidobacterium bifidum               |S83624||
| Bifidobacterium breve                |AB006658||
| Bifidobacterium longum                |AP010889| Bifidobacterium longum subsp. infa…|
| Brevibacterium massiliense            |EU086793 	|Brevibacterium ravenspurgense|
| Campylobacter ureolyticus             |ARGD01000016 |	Campylobacter ureolyticus DSM 20703|
| Corynebacterium amycolatum            |X84244||
| Corynebacterium aurimucosum           |AJ309207||
| Corynebacterium coyleae               |X96497||
| Corynebacterium matruchotii           |ACSH02000003 |	Corynebacterium matruchotii ATCC 1… | pintail was n/a|
| Corynebacterium pyruviciproducens     |ATBY01000001||
| Corynebacterium riegelii              |Y14651||
| Corynebacterium sp.                   |.|no type spec|
| Corynebacterium tuscaniense           |AY677186||
| Dermabacter hominis                   |X91034||
| Enterobacter asburiae                 |AB004744||
| Enterobacter cloacae                 |CP001918 	|Enterobacter cloacae subsp. cloaca…|
| Enterococcus faecalis                 |AB012212||
| Escherichia coli                      |EU014689||
| Facklamia hominis                     |AGZD01000007 	|Facklamia hominis CCUG 36813|pintail n/a
| Facklamia ignava                      |AGZE01000005 |	Facklamia ignava CCUG 37419
| Gardnerella vaginalis                 |M58744 |	Gardnerella vaginalis ATCC 14018|
| Globicatella sanguinis               | 	AB680901 |	Globicatella sanguinis|
| Gordonia terrae                       |AB920570||
| Klebsiella pneumoniae                |X87276||
| Kocuria rhizophila                    |Y16264||
| Kytococcus schroeteri                 |AJ297722||
| Lactobacillus crispatus               |AB008206||
| Lactobacillus delbrueckii             |AB007908||
| Lactobacillus fermentum               |AJ575812||
| Lactobacillus gasseri                 |AB008209||
| Lactobacillus iners                   |HE573916||
| Lactobacillus jensenii                |AF243176||
| Lactobacillus johnsonii                |HE573915||
| Lactobacillus pontis                   |X76329||
| Lactobacillus rhamnosus                |AB008211||
| Micrococcus luteus                     |CP001628 |	Micrococcus luteus NCTC 2665|
| Moraxella osloensis                  |FR726160||
| Morganella morganii                   | 	AJ301681 |	Morganella morganii subsp. morganii|
| Neisseria macacae                      |HF558383|
| Neisseria perflava                     |HF558366|
| Neisseria subflava                     |AJ239291|
| Oligella urethralis                    |AF133538|
| Propionibacterium acnes                |CP006032 |	Propionibacterium acnes hdn-1  no type spec
| Propionibacterium avidum              |AGBA01000019 |	as Cutibacterium avidum ATCC 25577|
| Proteus mirabilis                    |AJ301682||
| Pseudomonas aeruginosa                |HE978271||
| Rothia dentocariosa                   |CP002280 |	Rothia dentocariosa ATCC 17931|
| Rothia mucilaginosa                    |X87758||
| Staphylococcus epidermidis             |AB681292||
| Staphylococcus hominis                 |X66101 |	Staphylococcus hominis subsp. homi… 	1|
| Staphylococcus pettenkoferi           |AF322002||
| Staphylococcus saprophyticus           |AP008934 |	Staphylococcus saprophyticus subsp… 	|
| Staphylococcus simulans                |D83373||
| Staphylococcus warneri                |L37603||
| Streptococcus agalactiae               |AB023574||
| Streptococcus anginosus                |AFIM01000033||
| Streptococcus equinus                  |AB680295||
| Streptococcus gordonii                |AF003931||
| Streptococcus mitis                    |AF003929||
| Streptococcus oralis                  |ADMV01000001||
| Streptococcus parasanguinis            |AF003933||
| Streptococcus salivarius               |AY188352 	|Streptococcus salivarius subsp. sa… 	|
| Streptococcus sanguinis                |AF003928||
| Trueperella bernardiae                only one seq available |X79224||
| Varibaculum cambriense                 |AJ428402||
| Veillonella parvula                   |AB538437||

### pare down results

All these sequences were downloaded and saved in one multi-record FASTA file named arb-silva.de_2019-11-10_id739827.fasta. However, this file also includes several other entries of each species. For example, in addition to the desired type strain _Streptococcus equinus_ AB680295, Silva included an additional 6 entries in the download (AB002482.1.1457, AF104116.1.356, AEVB01000043.3380.4915, AF429764.1.1430, AF429765.1.1463 and AJ301607.1.1433). We used the script `src_files/cull_downloads.py` to comb through the Silva download and write the desired type strains to a new file `ktw_16s_type_2019-11-25_34.fna`"`.

```
~>python cull_downloads.py -i arb-silva.de_2019-11-10_id739827_tax_silva.fasta
writing to ktw_16s_type_2019-11-25_34.fna
```

## Alignment

The sequences in the `ktw_16s_type_2019-11-25_34.fna` file is then used to generate a multi-sequence alignement (MSA) of the 16S rRNA gene. We used the program T-Coffee[^5], an alignment program that generates accurate multisequence alignments even when the sequences are from unrelated species[^4]. The program was run with default settings.

[^4]: Edgar RC, Batzoglou S. Multiple sequence alignment. Current Opinion in Structural Biology. 2006 Jun;16(3):368–73. 
[^5]: Notredame C, Higgins DG, Heringa J. T-coffee: a novel method for fast and accurate multiple sequence alignment 1 1Edited by J. Thornton. Journal of Molecular Biology. 2000 Sep;302(1):205–17.

```
~$ t_coffee ktw_16s_type_2019-11-25_34.fna -outorder=aligned
```

The output of the program is a guide tree in Newick format (.dnd), an html file of the MSA in a colorized version, and the actual alignment in Clustal format (.aln). The Clustal file will be used to extract the predicted amplicons.

## Extract amplicons

The MSA contains the 16S rRNA sequence of the _E. coli_ type strain EU014689, and this will be used as a reference to map the 9 known variable regions onto the remaining 77 species from the Thomas-White dataset. The positions of the variable regions were confirmed through Sliding Window Analysis (see the sliding_window folder).

|variable region | _E. coli_ start | _E. coli_ stop | MSA start | MSA stop |
|---|---|---|---|---|
| V1 | 69 | 99|78  |151 |
| V2| 137| 242 |191 |333 |
| V3| 433 |497 |528 |596 |
| V4| 576| 682| 675 |783 |
| V5| 822| 879 |925 |986 |
| V6| 986| 1043|1093 |1163 |
| V7| 1117| 1173 | 1237 |1297 |
| V8| 1243| 1294 | 1369 |1424 |
| V9| 1435| 1465 | 1566 |1606 |


| paper | name | sequence | direction | E. coli pos | MSA pos | platform| notes |
|-----------|-------|--------|--------|----|----|----|--------------------|
|Komesu 2017| A17F |GTT TGA TCC TGG CTC AG|$\rightarrow$|12|19 | MiSeq| possible typo at pos 9 |
|Komesu 2017 | 515R|TTA CCG CGG CMG CSG GCA|$\leftarrow$|534| 633| MiSeq||
|Komesu 2017 |515F| GTG CCA GCT GCC GCG GTA ATA|$\rightarrow$|536|635 | MiSeq||
|Komesu 2017 |1114R|GGG GTT GCG CTC GTT GC|$\leftarrow$|1116|1234 | MiSeq||
|Bukin 2019| 16S_BV2f|AGT GGC GGA CGG GTG AGT AA|$\rightarrow$|102|154 | MiSeq||
|Bukin 2019|16S_BV3r| ... |$\leftarrow$|497?| 596| MiSeq|this sequence is exactly the same as BV2f, and isn't a palindrome. Typo? |
|Bukin 2019|MiCSQ_343FL|TAT GGT AAT TGT CTC CTA CGG RRS GCA GCA G|$\rightarrow$|344 | 435|MiSeq|no matches to MSA, assumed numbers are position of _E. coli_ 16S sequence[^1]|
|Bukin 2019|MiCSQ_806R|AGT CAG TCA GCC GGA CTA CNV GGG TWT CTA AT|$\leftarrow$|806 | 909|MiSeq|no matches to MSA, assumed numbers are position of _E. coli_ 16S sequence|
|Caporaso 2011|F515 |GTG CCA GCM GCC GCG GTA A | $\rightarrow$|516  | 616| pyro||
|Caporaso 2011|R806 |CCT GAT GHV CCC AWA GAT TA | $\leftarrow$| 808 | 911| pyro||
|Graspeuntner 2018|V3F|CCT ACG GGA GGC AGC AG|$\rightarrow$|342 | 434|MiSeq||
|Graspeuntner 2018|V4R|GGA CTA CHV GGG TWT CTA AT|$\leftarrow$|789 |892 |MiSeq||


| paper | name | sequence | direction | E. coli pos  | MSA pos |platform| notes |
|-----------|-------|--------|--------|----|----|--------------------|
|me | v3_579F|THT TSS RCA ATG GRS GVA|$\rightarrow$|359| 452 |MiSeq||
|me | v3_779R|GKN SCR AGC STT RHY CGG|$\leftarrow$|536| 635 |MiSeq||
|me | v6_1183F|CCG CCT GGG GAS TAC GVH|$\rightarrow$|875| 892 |MiSeq||
|me |v6_1410R|AGT CCC RYA ACG AGC GCA|$\leftarrow$|1083| 1203 |MiSeq||


![](../resources/md_files/amplicon_map.png)

```{r, fig.height=4, fig.width=8, echo=FALSE}
s16_70 <- read_tsv("../../sliding_window/processed_data/specific_windows_weighted_08_03_1426/ktw_16s_w70.csv", col_names = FALSE)
adj <- 35
vstart <- c(78,191,528,675,925,1093,1237,1369,1566)
vstop <- c(151,333,596,783,986,1163,1297,1424,1606)

pstart <- c(19,635,154,435,616,434,452,982)
pend <- c(633,1234,596,909,911,892,635,1203)
rank_pos <- c(.1, .2, .3, .4, .5, .6, .7, .8)

ggplot(s16_70, aes(x=X1, y=X2)) +
  geom_rect(xmin=vstart[1]-adj, xmax=vstop[1]-adj, ymin=-1, ymax=Inf, fill="ivory3") +
  geom_rect(xmin=vstart[2]-adj, xmax=vstop[2]-adj, ymin=-1, ymax=Inf, fill="ivory3") +
  geom_rect(xmin=vstart[3]-adj, xmax=vstop[3]-adj, ymin=-1, ymax=Inf, fill="ivory3") +
  geom_rect(xmin=vstart[4]-adj, xmax=vstop[4]-adj, ymin=-1, ymax=Inf, fill="ivory3") +
  geom_rect(xmin=vstart[5]-adj, xmax=vstop[5]-adj, ymin=-1, ymax=Inf, fill="ivory3") +
  geom_rect(xmin=vstart[6]-adj, xmax=vstop[6]-adj, ymin=-1, ymax=Inf, fill="ivory3") +
  geom_rect(xmin=vstart[7]-adj, xmax=vstop[7]-adj, ymin=-1, ymax=Inf, fill="ivory3") +
  geom_rect(xmin=vstart[8]-adj, xmax=vstop[8]-adj, ymin=-1, ymax=Inf, fill="ivory3") +
  geom_rect(xmin=vstart[9]-adj, xmax=vstop[9]-adj, ymin=-1, ymax=Inf, fill="ivory3") +
    annotate("text", x = 80, y = .95, label = "V1", color="gray40", size=5) +
  annotate("text", x = 230, y = .95, label = "V2", color="gray40", size=5) +
  annotate("text", x = 530, y = .95, label = "V3", color="gray40", size=5) +
  annotate("text", x = 700, y = .95, label = "V4", color="gray40", size=5) +
  annotate("text", x = 920, y = .95, label = "V5", color="gray40", size=5) +
  annotate("text", x = 1100, y = .95, label = "V6", color="gray40", size=5) +
  annotate("text", x = 1230, y = .95, label = "V7", color="gray40", size=5) +
  annotate("text", x = 1360, y = .95, label = "V8", color="gray40", size=5) +
  annotate("text", x = 1550, y = .95, label = "V9", color="gray40", size=5) +
   #bv
  geom_segment(aes(x = pstart[3], y = rank_pos[2], xend = pend[3], yend = rank_pos[2]), color = wes_palette("Darjeeling2")[5], size=2) +
  #geom_text(aes(label="16S_BV2f-16S_BV3r", x=1000, y=rank_pos[2]), size=4) +
  #geom_text(aes(label="Bukin 2019", x=950, y=rank_pos[2]), size=5) +
  geom_text(aes(label="V2-V3", x=700, y=rank_pos[2]), size=5) +
  
  #gras
  geom_segment(aes(x = pstart[6], y = rank_pos[5], xend = pend[6], yend = rank_pos[5]), color = wes_palette("Darjeeling2")[5], size=2) +
  #geom_text(aes(label="V3F-V4R", x=1200, y=rank_pos[5]), size=4) +
  #geom_text(aes(label="Graspeuntner 2018", x=1350, y=rank_pos[5]), size=5) +
  geom_text(aes(label="V3-V4", x=1000, y=rank_pos[5]), size=5) +
  
  #v3
  geom_segment(aes(x = pstart[7], y = rank_pos[3], xend = pend[7], yend = rank_pos[3]), color = wes_palette("Darjeeling2")[5], size=2) +
  #geom_text(aes(label="V3_579F-V3_779R", x=1070, y=rank_pos[3]), size=4) +
  geom_text(aes(label="V3", x=700, y=rank_pos[3]), size=5) +
  
  #cap
  geom_segment(aes(x = pstart[5], y = rank_pos[4], xend = pend[5], yend = rank_pos[4]), color = wes_palette("Darjeeling2")[5], size=2) +
  #geom_text(aes(label="F515-R806", x=1250, y=rank_pos[4]), size=4) +
  #geom_text(aes(label="Caporaso 2011", x=1300, y=rank_pos[4]), size=5) +
  geom_text(aes(label="V4", x=950, y=rank_pos[4]), size=5) +
  
  #k17
  geom_segment(aes(x = pstart[1], y = rank_pos[1], xend = pend[1], yend = rank_pos[1]), color = wes_palette("Darjeeling2")[5], size=2) +
  #geom_text(aes(label="A17F-515R", x=900, y=rank_pos[1]), size=4) +
  #geom_text(aes(label="Komesu 2017", x=1000, y=rank_pos[1]), size=5) +
  geom_text(aes(label="V1-V3", x=750, y=rank_pos[1]), size=5) +
  
  #b646
  geom_segment(aes(x = pstart[4], y = rank_pos[6], xend = pend[4], yend = rank_pos[6]), color = wes_palette("Darjeeling2")[5], size=2) +
  #geom_text(aes(label="MiCSQ_343FL-MiCSQ_806R", x=1400, y=rank_pos[6]), size=4) +
  #geom_text(aes(label="Bukin 2019", x=1250, y=rank_pos[6]), size=5) +
  geom_text(aes(label="V3-V5", x=1000, y=rank_pos[6]), size=5) +
  
  #k515
  geom_segment(aes(x = pstart[2], y = rank_pos[7], xend = pend[2], yend = rank_pos[7]), color = wes_palette("Darjeeling2")[5], size=2) +
  #geom_text(aes(label="515F-1114R", x=1600, y=rank_pos[7]), size=4) +
  #geom_text(aes(label="Komesu 2017", x=1650, y=rank_pos[7]), size=5) +
  geom_text(aes(label="V4-V6", x=1350, y=rank_pos[7]), size=5) +
  
  #v6
  geom_segment(aes(x = pstart[8], y = rank_pos[8], xend = pend[8], yend = rank_pos[8]), color = wes_palette("Darjeeling2")[5], size=2) +
  #geom_text(aes(label="V6_1183F-V6_1410R", x=1700, y=rank_pos[8]), size=4) +
  geom_text(aes(label="V6", x=1280, y=rank_pos[8]), size=5) +
  theme_classic() +
  theme(plot.title = element_text(size=20), plot.subtitle = element_text(size=14), axis.title.y= element_text(size=16), axis.title.x = element_text(size=16), axis.text = element_text(size=14), legend.title = element_text(size=14),legend.text = element_text(size=14), strip.text.x = element_text(size = 14), strip.text.y = element_text(size = 14)) +
  theme(axis.ticks.y = element_blank(), axis.text.y=element_blank()) +
  #labs(title="Position of amplicons on the 16S rRNA gene", x="Position in 16S gene sequence", y="Targeted amplicons")
  labs(x="Position in 16S gene sequence", y="Targeted amplicons")
```

The python script `src_files/extract_16s_vr.py` is used to extract the predicted amplicons from the MSA based on the 5' and 3' annealing sites of the primer set that would be used in PCR. In the example below, the script takes three arguments: the path to the MSA (ktw_type_16s_2019-11-25/aln/ktw_16s_type_2019-11-25_34.aln), the name of the file that will be written (cap_tpstr), and the start and stop coordinates (516,808). The output is a multi-record FASTA file of the predicted amplicon sequence, for each species in the MSA.

```
~>python extract_16s_vr.py -i ktw_type_16s_2019-11-25/aln/ktw_16s_type_2019-11-25_34.aln -o cap_tpstr -e 516,808
writing to ../processed_data/extracted_16s_regions_2019-11-26/extracted_cap_tpstr_2019-11-26_1550.fna
```

There are two important notes. The first is that you'll need to determine where a primer set's annealing sites correspond to the sequence of the _E. coli_ type strain EU014689. We used the program UGENE[^6] to search the MSA for the oligo sequence of the primer and transposed to _E. coli_, but there are other methods that will work. The second is that this script producs _predicted_ amplicons. It doesn't check if the primers will actually anneal to a specific species' 16S rRNA gene sequence. That's a feature that will probably be added later. 

[^6]: Okonechnikov K, Golosova O, Fursov M. Unipro UGENE: a unified bioinformatics toolkit. Bioinformatics. 2012 Apr;28(8):1166–7. 





