---
title: "for github"
author: "Carter Hoffman"
date: "_last edited `r Sys.Date()`_"

header-includes:
   - \usepackage{amsmath,amssymb}
   - \usepackage[bitstream-charter]{mathdesign}
   - \usepackage[T1]{fontenc}
   
output:
  md_document:
    variant: gfm

---
```{r, eval=FALSE, echo=FALSE}
output: 
  pdf_document:
      toc: true
      toc_depth: 2
      df_print: kable
    
output: 
  html_document:
    toc: true
    toc_depth: 4
    toc_float: true
    theme: readable
    highlight: tango

# knit directly to md for github
output:
  md_document:
    variant: gfm
```

```{r echo=FALSE}
knitr::opts_chunk$set(tidy = FALSE, warning = FALSE, message = FALSE, max.print = 5)
```

```{r echo=FALSE}
# load libraries
library(tidyverse)
library(wesanderson)
library(broom)
library(viridis)
library(ggrepel)
library(votesys)
```

# Overview

I'm using this file to generate all md files for the github readme pages.

![](resources/md_files/sm_overall_activity.png)

# Data download

I searched the Silva database on 11/10/2019 for each type strain of the Thomas-White dataset. Search parameters were the species name, SSU r132, and the strain=[T] designation. I chose the longest sequence length with the highest stats. For the species that had no hits, I used the synonum, if available.


| species |  type strain accession | notes|
|-----|-----|---------|
| Actinobaculum schaalii                | FJ960443 |as Actinotignum schaali|
| Actinomyces naeslundii                | AB618790||
| Actinomyces neuii                     | AM084229 |	Actinomyces neuii subsp. anitratus|
| Actinomyces odontolyticus             | AB818950||
| Actinomyces turicensis                | X78720||
| Actinomyces urogenitalis              | ACFH01000038 |	Actinomyces urogenitalis DSM 15434|
| Aerococcus christensenii              | Y17005||
| Aerococcus sanguinicola               | AJ276512||
| Aerococcus urinae                     | M77819||
| Alloscardovia omnicolens              | AM419460||
| Anaerococcus octavius                 | Y07841||
| Bacillus idriensis                    | AY904033||
| Bacillus infantis                     | AY904032||
| Bifidobacterium bifidum               | S83624||
| Bifidobacterium breve                 | AB006658||
| Bifidobacterium longum                | AP010889| Bifidobacterium longum subsp. infa…|
| Brevibacterium massiliense            | EU086793 	|Brevibacterium ravenspurgense|
| Campylobacter ureolyticus             | ARGD01000016 |	Campylobacter ureolyticus DSM 20703|
| Corynebacterium amycolatum            | X84244||
| Corynebacterium aurimucosum           | AJ309207||
| Corynebacterium coyleae               | X96497||
| Corynebacterium matruchotii           | ACSH02000003 |	Corynebacterium matruchotii ATCC 1… | pintail was n/a|
| Corynebacterium pyruviciproducens     | ATBY01000001||
| Corynebacterium riegelii              | Y14651||
| Corynebacterium sp.                   | .|no type spec|
| Corynebacterium tuscaniense           | AY677186||
| Dermabacter hominis                   | X91034||
| Enterobacter asburiae                 | AB004744||
| Enterobacter cloacae                  | CP001918 	|Enterobacter cloacae subsp. cloaca…|
| Enterococcus faecalis                 | AB012212||
| Escherichia coli                      | EU014689||
| Facklamia hominis                     | AGZD01000007 	|Facklamia hominis CCUG 36813|pintail n/a
| Facklamia ignava                      | AGZE01000005 |	Facklamia ignava CCUG 37419
| Gardnerella vaginalis                 | M58744 |	Gardnerella vaginalis ATCC 14018|
| Globicatella sanguinis                | AB680901 |	Globicatella sanguinis|
| Gordonia terrae                       | AB920570||
| Klebsiella pneumoniae                 | X87276||
| Kocuria rhizophila                    | Y16264||
| Kytococcus schroeteri                 | AJ297722||
| Lactobacillus crispatus               | AB008206||
| Lactobacillus delbrueckii             | AB007908||
| Lactobacillus fermentum               | AJ575812||
| Lactobacillus gasseri                 | AB008209||
| Lactobacillus iners                   | HE573916||
| Lactobacillus jensenii                | AF243176||
| Lactobacillus johnsonii               | HE573915||
| Lactobacillus pontis                  | X76329||
| Lactobacillus rhamnosus               | AB008211||
| Micrococcus luteus                    | CP001628 |	Micrococcus luteus NCTC 2665|
| Moraxella osloensis                   | FR726160||
| Morganella morganii                   | AJ301681 |	Morganella morganii subsp. morganii|
| Neisseria macacae                     | HF558383|
| Neisseria perflava                    | HF558366|
| Neisseria subflava                    | AJ239291|
| Oligella urethralis                   | AF133538|
| Propionibacterium acnes               | CP006032 |	Propionibacterium acnes hdn-1  no type spec
| Propionibacterium avidum              | AGBA01000019 |	as Cutibacterium avidum ATCC 25577|
| Proteus mirabilis                     | AJ301682||
| Pseudomonas aeruginosa                | HE978271||
| Rothia dentocariosa                   | CP002280 |	Rothia dentocariosa ATCC 17931|
| Rothia mucilaginosa                   | X87758||
| Staphylococcus epidermidis            | AB681292||
| Staphylococcus hominis                | X66101 |	Staphylococcus hominis subsp. homi… 	1|
| Staphylococcus pettenkoferi           | AF322002||
| Staphylococcus saprophyticus          | AP008934 |	Staphylococcus saprophyticus subsp… 	|
| Staphylococcus simulans               | D83373||
| Staphylococcus warneri                | L37603||
| Streptococcus agalactiae              | AB023574||
| Streptococcus anginosus               | AFIM01000033||
| Streptococcus equinus                 | AB680295||
| Streptococcus gordonii                | AF003931||
| Streptococcus mitis                   | AF003929||
| Streptococcus oralis                  | ADMV01000001||
| Streptococcus parasanguinis           | AF003933||
| Streptococcus salivarius              | AY188352 	|Streptococcus salivarius subsp. sa… 	|
| Streptococcus sanguinis               | AF003928||
| Trueperella bernardiae                | X79224||
| Varibaculum cambriense                | AJ428402||
| Veillonella parvula                   | AB538437||

The files were downloaded and saved in folder arb-silva.de_2019-11-10_id739827. A copy of the file was made and renamed ktw_16s_type_strains_2019-11-10.

# data prep

T-Coffee requires that the FASTA description line only contains the ID. I wrote a script that selects only the accession numbers from the sequences and writes to a new file. The script is in /thesis/gather_markers/ssu16s/bladder_population/arb-silva.de_2019-11-10_id739827/cull_downloads.py, so don't look for it in /thesis_git. The basic function is:

```{r, eval=FALSE}
with open("accessions.txt", 'r') as f:
    acc_info=f.readlines()
    
accessions=[x.split('|')[4].strip() for x in acc_info]
get_records={x.id.split('.')[0]:x for x in SeqIO.parse("arb-silva.de_2019-11-10_id739827_tax_silva.fasta", "fasta")}
get_ids=[l for l in get_records.keys()]
smaller=[g for g in get_ids if g in accessions]

with open("ktw_16s_type_{}.fna".format(runtime), 'a') as o:
    for s in smaller:
        SeqIO.write(get_records[s], o, 'fasta')
```

# alignment

Then ran it through tcoffee.

```{shellsession, eval=FALSE}
Microsoft Windows [Version 10.0.18362.476]
(c) 2019 Microsoft Corporation. All rights reserved.

C:\Users\Carter>ssh hoffmanc@acc.ohsu.edu

Success. Logging you in...
Last login: Mon Nov 25 18:58:01 2019 from 65.182.242.40
Success. Logging you in...
Last login: Mon Nov 25 18:58:01 2019 from 65.182.242.40
-bash-4.3$ ssh hoffmanc@state
hoffmanc@states password:
Welcome to Ubuntu 16.04.6 LTS (GNU/Linux 4.4.0-116-generic x86_64)


Last login: Sat Nov  9 16:13:13 2019 from 10.96.10.55

hoffmanc@state:~$ mv ktw_16s_type_2019-11-25_34.fna tw_markergenes/

hoffmanc@state:~$ cd tw_markergenes/

hoffmanc@state:~/tw_markergenes$ cd ../tcoffee/

hoffmanc@state:~/tcoffee$ cd Version_12.00.7fb08c2/

hoffmanc@state:~$ cd tw_markergenes/coffee_output/
hoffmanc@state:~/tw_markergenes/coffee_output$ mv ../ktw_16s_type_2019-11-25_34.fna ./

hoffmanc@state:~/tw_markergenes/coffee_output$ t_coffee ktw_16s_type_2019-11-25_34.fna -outorder=aligned

PROGRAM: T-COFFEE Version_12.00.7fb08c2 (2018-12-11 09:27:12 - Revision 7fb08c2 - Build 211)
-full_log       S       [0]
-genepred_score S       [0]     nsd
-run_name       S       [0]
-mem_mode       S       [0]     mem
-extend         D       [1]     1
-extend_mode    S       [0]     very_fast_triplet
-max_n_pair     D       [0]     10
-seq_name_for_quadruplet        S       [0]     all
-compact        S       [0]     default
-clean          S       [0]     no
-do_self        FL      [0]     0
-do_normalise   D       [0]     1000
-template_file  S       [0]
-setenv         S       [0]     0
-export         S       [0]     0
-template_mode  S       [0]
-flip           D       [0]     0
-remove_template_file   D       [0]     0
-profile_template_file  S       [0]
-in             S       [0]
-seq            S       [1]     ktw_16s_type_2019-11-25_34.fna
-aln            S       [0]
-method_limits  S       [0]
-method         S       [0]
-lib            S       [0]
-profile        S       [0]
-profile1       S       [0]
-profile2       S       [0]
-pdb            S       [0]
-relax_lib      D       [0]     1
-filter_lib     D       [0]     0
-shrink_lib     D       [0]     0
-out_lib        W_F     [0]     no
-out_lib_mode   S       [0]     primary
-lib_only       D       [0]     0
-outseqweight   W_F     [0]     no
-seq_source     S       [0]     ANY
-cosmetic_penalty       D       [0]     0
-gapopen        D       [0]     0
-gapext         D       [0]     0
-fgapopen       D       [0]     0
-fgapext        D       [0]     0
-nomatch        D       [0]     0
-newtree        W_F     [0]     default
-tree           W_F     [0]     NO
-usetree        R_F     [0]
-tree_mode      S       [0]     nj
-distance_matrix_mode   S       [0]     ktup
-distance_matrix_sim_mode       S       [0]     idmat_sim1
-quicktree      FL      [0]     0
-outfile        W_F     [0]     default
-maximise       FL      [1]     1
-output         S       [0]     aln     html
-len            D       [0]     0
-infile         R_F     [0]
-matrix         S       [0]     default
-tg_mode        D       [0]     1
-profile_mode   S       [0]     cw_profile_profile
-profile_comparison     S       [0]     profile
-dp_mode        S       [0]     linked_pair_wise
-ktuple         D       [0]     1
-ndiag          D       [0]     0
-diag_threshold D       [0]     0
-diag_mode      D       [0]     0
-sim_matrix     S       [0]     vasiliky
-transform      S       [0]
-extend_seq     FL      [0]     0
-outorder       S       [1]     aligned
-inorder        S       [0]     aligned
-seqnos         S       [0]     off
-case           S       [0]     keep
-cpu            D       [0]     0
-ulimit         D       [0]     -1
-maxnseq        D       [0]     -1
-maxlen         D       [0]     -1
-sample_dp      D       [0]     0
-weight         S       [0]     default
-seq_weight     S       [0]     no
-align          FL      [1]     1
-mocca          FL      [0]     0
-domain         FL      [0]     0
-start          D       [0]     0
-len            D       [0]     0
-scale          D       [0]     0
-mocca_interactive      FL      [0]     0
-method_evaluate_mode   S       [0]     default
-color_mode     S       [0]     new
-aln_line_length        D       [0]     0
-evaluate_mode  S       [0]     triplet
-get_type       FL      [0]     0
-clean_aln      D       [0]     0
-clean_threshold        D       [1]     1
-clean_iteration        D       [1]     1
-clean_evaluate_mode    S       [0]     t_coffee_fast
-extend_matrix  FL      [0]     0
-prot_min_sim   D       [40]    40
-prot_max_sim   D       [90]    90
-prot_trim      D       [20]    20
-prot_min_cov   D       [40]    40
-pdb_type       S       [0]     d
-pdb_min_sim    D       [35]    35
-pdb_max_sim    D       [100]   100
-pdb_min_cov    D       [50]    50
-pdb_blast_server       W_F     [0]     EBI
-blast          W_F     [0]
-blast_server   W_F     [0]     EBI
-pdb_db         W_F     [0]     pdb
-protein_db     W_F     [0]     uniref50
-method_log     W_F     [0]     no
-struc_to_use   S       [0]
-cache          W_F     [0]     use
-print_cache    FL      [0]     0
-align_pdb_param_file   W_F     [0]     no
-align_pdb_hasch_mode   W_F     [0]     hasch_ca_trace_bubble
-external_aligner       S       [0]     NO
-msa_mode       S       [0]     tree
-et_mode        S       [0]     et
-master         S       [0]     no
-blast_nseq     D       [0]     0
-lalign_n_top   D       [0]     10
-iterate        D       [0]     0
-trim           D       [0]     0
-split          D       [0]     0
-trimfile       S       [0]     default
-split          D       [0]     0
-split_nseq_thres       D       [0]     0
-split_score_thres      D       [0]     0
-check_pdb_status       D       [0]     0
-clean_seq_name D       [0]     0
-seq_to_keep    S       [0]
-dpa_master_aln S       [0]
-dpa_maxnseq    D       [0]     0
-dpa_min_score1 D       [0]
-dpa_min_score2 D       [0]
-dpa_keep_tmpfile       FL      [0]     0
-dpa_debug      D       [0]     0
-multi_core     S       [0]     templates_jobs_relax_msa_evaluate
-n_core         D       [0]     0
-max_n_proc     D       [0]     0
-lib_list       S       [0]
-prune_lib_mode S       [0]     5
-tip            S       [0]     none
-rna_lib        S       [0]
-no_warning     D       [0]     0
-run_local_script       D       [0]     0
-plugins        S       [0]     default
-proxy          S       [0]     unset
-email          S       [0]
-clean_overaln  D       [0]     0
-overaln_param  S       [0]
-overaln_mode   S       [0]
-overaln_model  S       [0]
-overaln_threshold      D       [0]     0
-overaln_target D       [0]     0
-overaln_P1     D       [0]     0
-overaln_P2     D       [0]     0
-overaln_P3     D       [0]     0
-overaln_P4     D       [0]     0
-exon_boundaries        S       [0]
-dump           S       [0]     no
-display        D       [0]     100

INPUT FILES
        Input File (S) ktw_16s_type_2019-11-25_34.fna  Format fasta_seq
        Input File (M) proba_pair

Identify Master Sequences [no]:

Master Sequences Identified
INPUT SEQUENCES: 78 SEQUENCES  [RNA]
        Multi Core Mode: 48 processors:

        --- Process Method/Library/Aln Sktw_16s_type_2019-11-25_34.fna
        --- Process Method/Library/Aln Mproba_pair
        xxx Retrieved Sktw_16s_type_2019-11-25_34.fna
        xxx Retrieved Mproba_pair

        All Methods Retrieved

MANUAL PENALTIES: gapopen=0 gapext=0

        Library Total Size: [13803828]

Library Relaxation: Multi_proc [48]

!               [Relax Library][TOT=    1][100 %]
Relaxation Summary: [13803828]--->[12610885]



UN-WEIGHTED MODE: EVERY SEQUENCE WEIGHTS 1

MAKE GUIDE TREE
        [MODE=nj][DONE]

PROGRESSIVE_ALIGNMENT [Tree Based]
Group    1: AB004744.1.1422
Group    2: AB006658.1.1520
Group    3: AB007908.1.1516
Group    4: AB008206.1.1556
Group    5: AB008209.1.1566
Group    6: AB008211.1.1540
Group    7: AB012212.1.1517
Group    8: AB023574.1.1501
Group    9: AB538437.1.1524
Group   10: AB618790.1.1522
Group   11: AB680295.1.1467
Group   12: AB680901.1.1475
Group   13: AB681292.1.1476
Group   14: AB818950.1.1501
Group   15: AB920570.1.1436
Group   16: ACFH01000038.241.1767
Group   17: ACSH02000003.434.1958
Group   18: ADMV01000001.3349.4885
Group   19: AF003928.1.1460
Group   20: AF003929.1.1520
Group   21: AF003931.1.1512
Group   22: AF003933.1.1513
Group   23: AF133538.1.1486
Group   24: AF243176.1.1496
Group   25: AF322002.1.1547
Group   26: AFIM01000033.206.1753
Group   27: AGBA01000019.373.1883
Group   28: AGZD01000007.268077.269614
Group   29: AGZE01000005.9117.10656
Group   30: AJ239291.1.1355
Group   31: AJ276512.1.1499
Group   32: AJ297722.1.1475
Group   33: AJ301681.1.1502
Group   34: AJ301682.1.1503
Group   35: AJ309207.1.1475
Group   36: AJ428402.1.1510
Group   37: AJ575812.1.1555
Group   38: AM084229.1.1497
Group   39: AM419460.1.1482
Group   40: AP008934.2304649.2306203
Group   41: AP010889.960839.962368
Group   42: ARGD01000016.303.1794
Group   43: ATBY01000001.60.1575
Group   44: AY188352.1.1546
Group   45: AY677186.1.1502
Group   46: AY904032.1.1367
Group   47: AY904033.1.1437
Group   48: CP001628.419927.421558
Group   49: CP001918.5152784.5154334
Group   50: CP002280.1382146.1383660
Group   51: CP006032.597533.599065
Group   52: D83373.1.1476
Group   53: EU014689.1.1541
Group   54: EU086793.1.1470
Group   55: FJ960443.1.1495
Group   56: FR726160.1.1591
Group   57: HE573915.1.1543
Group   58: HE573916.1.1560
Group   59: HE978271.1.1527
Group   60: HF558366.1.1514
Group   61: HF558383.1.1514
Group   62: L37603.1.1470
Group   63: M58744.1.1475
Group   64: M77819.1.1481
Group   65: S83624.1.1532
Group   66: X66101.1.1544
Group   67: X76329.1.1570
Group   68: X78720.1.1453
Group   69: X79224.1.1382
Group   70: X84244.1.1476
Group   71: X87276.1.1534
Group   72: X87758.1.1467
Group   73: X91034.1.1511
Group   74: X96497.1.1483
Group   75: Y07841.1.1407
Group   76: Y14651.1.1383
Group   77: Y16264.1.1471
Group   78: Y17005.1.1451

        Group  155: [Group   42 (   1 seq)] with [Group  154 (  77 seq)]-->[Len= 1835][PID:112737]


!               [Final Evaluation][TOT=   38][100 %]



OUTPUT RESULTS
        #### File Type= GUIDE_TREE      Format= newick          Name= ktw_16s_type_2019-11-25_34.dnd
        #### File Type= MSA             Format= aln             Name= ktw_16s_type_2019-11-25_34.aln
        #### File Type= MSA             Format= html            Name= ktw_16s_type_2019-11-25_34.html

# Command Line: t_coffee ktw_16s_type_2019-11-25_34.fna -outorder aligned  [PROGRAM:T-COFFEE]
# T-COFFEE Memory Usage: Current= 41.155 Mb, Max= 305.010 Mb
# Results Produced with T-COFFEE Version_12.00.7fb08c2 (2018-12-11 09:27:12 - Revision 7fb08c2 - Build 211)
# T-COFFEE is available from http://www.tcoffee.org
# Register on: https://groups.google.com/group/tcoffee/
```

The T-Coffee output was zipped up and downloaded.


# split to vr

Using the _E. coli_ type strain EU014689.1.1541 for reference coordinates

|variable region | _E. coli_ start | _E. coli_ stop |
|---|---|---|---|---|
| V1 | 69 | 99|
| V2| 137| 242 |
| V3| 433 |497 |
| V4| 576| 682| 
| V5| 822| 879 |
| V6| 986| 1043|
| V7| 1117| 1173 | 
| V8| 1243| 1294 | 
| V9| 1435| 1465 | 

##### published primers

| paper | name | sequence | direction | E. coli pos |platform| notes |
|-----------|-------|--------|--------|----|----|--------------------|
|Komesu 2017| A17F |GTT TGA TCC TGG CTC AG|$\rightarrow$|12|MiSeq| possible typo at pos 9 |
|Komesu 2017 | 515R|TTA CCG CGG CMG CSG GCA|$\leftarrow$|534|MiSeq||
|Komesu 2017 |515F| GTG CCA GCT GCC GCG GTA ATA|$\rightarrow$|536|MiSeq||
|Komesu 2017 |1114R|GGG GTT GCG CTC GTT GC|$\leftarrow$|116|MiSeq||
|Bukin 2019| 16S_BV2f|AGT GGC GGA CGG GTG AGT AA|$\rightarrow$|102|MiSeq||
|Bukin 2019|16S_BV3r| ... |$\leftarrow$|497?|MiSeq|this sequence is exactly the same as BV2f, and isn't a palindrome. Typo? |
|Bukin 2019|MiCSQ_343FL|TAT GGT AAT TGT CTC CTA CGG RRS GCA GCA G|$\rightarrow$|344|MiSeq|no matches to MSA, assumed numbers are position of _E. coli_ 16S sequence[^1]|
|Bukin 2019|MiCSQ_806R|AGT CAG TCA GCC GGA CTA CNV GGG TWT CTA AT|$\leftarrow$|806|MiSeq|no matches to MSA, assumed numbers are position of _E. coli_ 16S sequence|
|Caporaso 2011|F515 |GTG CCA GCM GCC GCG GTA A | $\rightarrow$|516 | pyro||
|Caporaso 2011|R806 |CCT GAT GHV CCC AWA GAT TA | $\leftarrow$| 808| pyro||
|Graspeuntner 2018|V3F|CCT ACG GGA GGC AGC AG|$\rightarrow$|342|MiSeq||
|Graspeuntner 2018|V4R|GGA CTA CHV GGG TWT CTA AT|$\leftarrow$|789|MiSeq||

##### primers I designed

| paper | name | sequence | direction | E. coli pos  |platform| notes |
|-----------|-------|--------|--------|----|----|--------------------|
|me | v3_579F|THT TSS RCA ATG GRS GVA|$\rightarrow$|359|MiSeq||
|me | v3_779R|GKN SCR AGC STT RHY CGG|$\leftarrow$|536|MiSeq||
|me | v6_1183F|CCG CCT GGG GAS TAC GVH|$\rightarrow$|875|MiSeq||
|me |v6_1410R|AGT CCC RYA ACG AGC GCA|$\leftarrow$|1083|MiSeq||

### extraction

Primers and the abbreviations I'm using.

| paper | primer | abbreviation |
|-----------|-------|--------|
|Komesu 2017| A17F-515R|k17|
|Komesu 2017 |515F-1114R|k515|
|Bukin 2019| 16S_BV2f-16S_BV3r|bbv|
|Bukin 2019|MiCSQ_343FL-MiCSQ_806R|b646|
|Caporaso 2011|F515-R806|cap|
|Graspeuntner 2018|V3F-V4R|gras|
|me | v6_1183F-v6_1410R|v6|
|me | v3_579F-v3_779R|v3|


Each primer set coordinates were used to extract the subsequences from the multisequence alignment. 

```{shellsession, eval=FALSE}
(base) C:\Users\Carter\Documents\thesis\thesis_git\split_to_vr\src_files>python extract_16s_vr.py -i ../../resources/ktw_type_16s_2019-11-25/aln/ktw_16s_type_2019-11-25_34.aln -o cap_tpstr -e 516,808
writing to ../processed_data/extracted_16s_regions_2019-11-26/extracted_cap_tpstr_2019-11-26_1550.fna

(base) C:\Users\Carter\Documents\thesis\thesis_git\split_to_vr\src_files>python extract_16s_vr.py -i ../../resources/ktw_type_16s_2019-11-25/aln/ktw_16s_type_2019-11-25_34.aln -o gras_tpstr -e 342,789
writing to ../processed_data/extracted_16s_regions_2019-11-26/extracted_gras_tpstr_2019-11-26_1551.fna

(base) C:\Users\Carter\Documents\thesis\thesis_git\split_to_vr\src_files>python extract_16s_vr.py -i ../../resources/ktw_type_16s_2019-11-25/aln/ktw_16s_type_2019-11-25_34.aln -o k17_tpstr -e 12,534
writing to ../processed_data/extracted_16s_regions_2019-11-26/extracted_k17_tpstr_2019-11-26_1551.fna

(base) C:\Users\Carter\Documents\thesis\thesis_git\split_to_vr\src_files>python extract_16s_vr.py -i ../../resources/ktw_type_16s_2019-11-25/aln/ktw_16s_type_2019-11-25_34.aln -o k515_tpstr -e 536,1116
writing to ../processed_data/extracted_16s_regions_2019-11-26/extracted_k515_tpstr_2019-11-26_1551.fna

(base) C:\Users\Carter\Documents\thesis\thesis_git\split_to_vr\src_files>python extract_16s_vr.py -i ../../resources/ktw_type_16s_2019-11-25/aln/ktw_16s_type_2019-11-25_34.aln -o bbv_tpstr -e 102,497
writing to ../processed_data/extracted_16s_regions_2019-11-26/extracted_bbv_tpstr_2019-11-26_1552.fna

(base) C:\Users\Carter\Documents\thesis\thesis_git\split_to_vr\src_files>python extract_16s_vr.py -i ../../resources/ktw_type_16s_2019-11-25/aln/ktw_16s_type_2019-11-25_34.aln -o b646_tpstr -e 344,806
writing to ../processed_data/extracted_16s_regions_2019-11-26/extracted_b646_tpstr_2019-11-26_1552.fna

(base) C:\Users\Carter\Documents\thesis\thesis_git\split_to_vr\src_files>python extract_16s_vr.py -i ../../resources/ktw_type_16s_2019-11-25/aln/ktw_16s_type_2019-11-25_34.aln -o v3_tpstr -e 359,536
writing to ../processed_data/extracted_16s_regions_2019-11-26/extracted_v3_tpstr_2019-11-26_1552.fna

(base) C:\Users\Carter\Documents\thesis\thesis_git\split_to_vr\src_files>python extract_16s_vr.py -i ../../resources/ktw_type_16s_2019-11-25/aln/ktw_16s_type_2019-11-25_34.aln -o v6_tpstr -e 875,1083
writing to ../processed_data/extracted_16s_regions_2019-11-26/extracted_v6_tpstr_2019-11-26_1552.fna
```

# classify

## blca

I put all the BLCA commands for every database into a bash script. Run all at once.

Moved the extracted_16s_regions_2019-11-26 dir and blca_typestr_alldb.bash to the Ubuntu side, ran it.

```{shellsession, eval=FALSE}
buzz@costco-special-k0Mput3r:~/BLCA$ chmod +x blca_typestr_alldb.bash
buzz@costco-special-k0Mput3r:~/BLCA$ ./blca_typestr_alldb.bash
```

ran without errors

### reformatting

```{shellsession, eval=FALSE}
(base) C:\Users\Carter\Documents\thesis\thesis_git\taxonomy\src_files>python blcaout_to_csv.py -i ../raw_data/blca_tpstr_2019-11-26
writing to ../processed_files/formatted_outfiles_2019-11-26_2031
writing to ../processed_files/formatted_outfiles_2019-11-26_2031/formatted_b646_tpstr_default_2019-11-26_2019-11-26_2031.csv
writing to ../processed_files/formatted_outfiles_2019-11-26_2031/formatted_b646_tpstr_genomic_2019-11-26_2019-11-26_2031.csv
writing to ../processed_files/formatted_outfiles_2019-11-26_2031/formatted_b646_tpstr_gg_2019-11-26_2019-11-26_2031.csv
writing to ../processed_files/formatted_outfiles_2019-11-26_2031/formatted_b646_tpstr_silva_2019-11-26_2019-11-26_2031.csv
writing to ../processed_files/formatted_outfiles_2019-11-26_2031/formatted_bbv_tpstr_default_2019-11-26_2019-11-26_2031.csv
writing to ../processed_files/formatted_outfiles_2019-11-26_2031/formatted_bbv_tpstr_genomic_2019-11-26_2019-11-26_2031.csv
writing to ../processed_files/formatted_outfiles_2019-11-26_2031/formatted_bbv_tpstr_gg_2019-11-26_2019-11-26_2031.csv
writing to ../processed_files/formatted_outfiles_2019-11-26_2031/formatted_bbv_tpstr_silva_2019-11-26_2019-11-26_2031.csv
writing to ../processed_files/formatted_outfiles_2019-11-26_2031/formatted_cap_tpstr_default_2019-11-26_2019-11-26_2031.csv
writing to ../processed_files/formatted_outfiles_2019-11-26_2031/formatted_cap_tpstr_genomic_2019-11-26_2019-11-26_2031.csv
writing to ../processed_files/formatted_outfiles_2019-11-26_2031/formatted_cap_tpstr_gg_2019-11-26_2019-11-26_2031.csv
writing to ../processed_files/formatted_outfiles_2019-11-26_2031/formatted_cap_tpstr_silva_2019-11-26_2019-11-26_2031.csv
writing to ../processed_files/formatted_outfiles_2019-11-26_2031/formatted_gras_tpstr_default_2019-11-26_2019-11-26_2031.csv
writing to ../processed_files/formatted_outfiles_2019-11-26_2031/formatted_gras_tpstr_genomic_2019-11-26_2019-11-26_2031.csv
writing to ../processed_files/formatted_outfiles_2019-11-26_2031/formatted_gras_tpstr_gg_2019-11-26_2019-11-26_2031.csv
writing to ../processed_files/formatted_outfiles_2019-11-26_2031/formatted_gras_tpstr_silva_2019-11-26_2019-11-26_2031.csv
writing to ../processed_files/formatted_outfiles_2019-11-26_2031/formatted_k17_tpstr_default_2019-11-26_2019-11-26_2031.csv
writing to ../processed_files/formatted_outfiles_2019-11-26_2031/formatted_k17_tpstr_genomic_2019-11-26_2019-11-26_2031.csv
writing to ../processed_files/formatted_outfiles_2019-11-26_2031/formatted_k17_tpstr_gg_2019-11-26_2019-11-26_2031.csv
writing to ../processed_files/formatted_outfiles_2019-11-26_2031/formatted_k17_tpstr_silva_2019-11-26_2019-11-26_2031.csv
writing to ../processed_files/formatted_outfiles_2019-11-26_2031/formatted_k515_tpstr_default_2019-11-26_2019-11-26_2031.csv
writing to ../processed_files/formatted_outfiles_2019-11-26_2031/formatted_k515_tpstr_genomic_2019-11-26_2019-11-26_2031.csv
writing to ../processed_files/formatted_outfiles_2019-11-26_2031/formatted_k515_tpstr_gg_2019-11-26_2019-11-26_2031.csv
writing to ../processed_files/formatted_outfiles_2019-11-26_2031/formatted_k515_tpstr_silva_2019-11-26_2019-11-26_2031.csv
writing to ../processed_files/formatted_outfiles_2019-11-26_2031/formatted_v3_tpstr_default_2019-11-26_2019-11-26_2031.csv
writing to ../processed_files/formatted_outfiles_2019-11-26_2031/formatted_v3_tpstr_genomic_2019-11-26_2019-11-26_2031.csv
writing to ../processed_files/formatted_outfiles_2019-11-26_2031/formatted_v3_tpstr_gg_2019-11-26_2019-11-26_2031.csv
writing to ../processed_files/formatted_outfiles_2019-11-26_2031/formatted_v3_tpstr_silva_2019-11-26_2019-11-26_2031.csv
writing to ../processed_files/formatted_outfiles_2019-11-26_2031/formatted_v6_tpstr_default_2019-11-26_2019-11-26_2031.csv
writing to ../processed_files/formatted_outfiles_2019-11-26_2031/formatted_v6_tpstr_genomic_2019-11-26_2019-11-26_2031.csv
writing to ../processed_files/formatted_outfiles_2019-11-26_2031/formatted_v6_tpstr_gg_2019-11-26_2019-11-26_2031.csv
writing to ../processed_files/formatted_outfiles_2019-11-26_2031/formatted_v6_tpstr_silva_2019-11-26_2019-11-26_2031.csv
```

Next I had to make a map file for the type strain accesion numbers and known taxonomy.

### typo!

I just noticed a typo. `b646` should be `b343`. Leaving it for now, but beware.

```{shellsession, eval=FALSE}
(base) C:\Users\Carter\Documents\thesis\thesis_git\taxonomy\src_files>python map_id_tax.py -h
usage: map_id_tax.py [-h] [--infile INFILE] [--outfile OUTFILE] [--verbose]

    Record descriptions of fasta files are lost along the pipeline.
    Use this script to make a csv file that maps the accession number to the species taxon.

    Changes the fasta record description line from this:

        CP006690.1.778390 Bacteria;Firmicutes;Bacilli;Lactobacillales;Lactobacillaceae;Lactobacillus;Lactobacillus casei 12A

    to a csv formatted like this:

        CP006690.1.778390,Lactobacillus casei 12A

    file is automatically saved in /resources

optional arguments:
  -h, --help            show this help message and exit
  --infile INFILE, -i INFILE
                        The full or relative path to the file.
  --outfile OUTFILE, -o OUTFILE
                        The name of the new file to be written
  --verbose, -v         spell out what is happening

This script doesnt have any other dependencies besides biopython

(base) C:\Users\Carter\Documents\thesis\thesis_git\taxonomy\src_files>python map_id_tax.py -i ../../resources/ktw_16s_type_2019-11-25_34.fna -o type_strains_map
```

There were a couple lines where the script didn't pull out the names correctly, like `X96497.1.1483,Corynebacterium 1;Corynebacterium coyleae` so cleaned that up by hand. Then ran through the synonyms. I would like to say that nothing will happen, because I checked for synonyms when I donwnloaded the type strains, but we'll see.

## naive bayes

fill in here

# synonyms

`validate_match.py` compares the results of the taxonomy assigned by a classifier to the known taxonomy held in a map file.

## DSMZ

Prokaryotic synonyms were downloaded from the [Prokaryotic Nomenclature Up-to-Date website](https://www.dsmz.de/services/online-tools/prokaryotic-nomenclature-up-to-date/downloads) run by the DSMZ-German Collection of Microorganisms and Cell Cultures, GmbH. If the synonym names are already the same, it's because the DSMZ included information down to the strain level. Strains are out of scope, so entries like _Enterobacter cloacae cloacae_ and _Enterobacter cloacae dissolvens_ are considered synonyms of _Enterobacter cloacae_. Misspellings that don't map to known synonyms are assigned as a mismatch. I'll fix that later, but know that unless you're typing your own taxonomy into fasta records, the chance of misspellings are low.

## in general

Any genus or species taxon labled as "unidentified" or "uncultured" is designated as "unavailable". "Unclassified" is reserved for when the classification algo thinks there's no match for the record. "Unavailable" just means there's no information for that record.

Input is the formatted csv file as written by the python script 'blcaout_to_csv.py'

```
id,taxon,rank,confidence
CP011325.22226.23780,Bacteria,domain,100.0
```

the --map_file option takes a path to a file that can be used to map accession numbers to taxonomy by newick_tools.inhouse_to_species(). See below for the list of some existing map files. 
run the script like this:

```
>python validate_match.py -i path/to/formatted/assigned_taxonomy_file.csv -m path/to/map/file
```

# evaluations






